src/aliceVision/track/TracksBuilder.cpp

```diff
@@ -52,7 +73,99 @@ TracksBuilder::TracksBuilder()
 
 TracksBuilder::~TracksBuilder() = default;
 
-void TracksBuilder::build(const PairwiseMatches& pairwiseMatches)
+void mergeTracks(const feature::MapFeaturesPerView& featuresPerView, const MapIndexToNode& map_indexToNode,
+                 const PairwiseMatches& pairwiseMatches, std::unique_ptr<TracksBuilderData>& _d,
+                 stl::flat_map<IndexedFeaturePair, size_t>& _duplicateFeaturesMap)
+{
+    // make the union according to duplicate features
+    // (same position, scale and describer type, but different orientations)
+
+    // map of (viewId) to
+    //    map of (descType) to
+    //        map of DuplicateFeatureId(x, y, scale) to
+    //            pair of (set<featureId>, node)
+    HashMap<size_t, HashMap<feature::EImageDescriberType,
+                    HashMap<DuplicateFeatureId, std::pair<std::set<size_t>, const MapIndexToNode::mapped_type*>>>>
+        duplicateFeaturesPerView;
+
+    // per viewId pair
+    for(const auto& matchesPerDescIt : pairwiseMatches)
+    {
+        const std::size_t& I = matchesPerDescIt.first.first;
+        const std::size_t& J = matchesPerDescIt.first.second;
+        const MatchesPerDescType& matchesPerDesc = matchesPerDescIt.second;
+
+        auto& featuresPerDescI = featuresPerView.at(I);
+        auto& featuresPerDescJ = featuresPerView.at(J);
+        auto& duplicateFeaturesPerDescI = duplicateFeaturesPerView[I];
+        auto& duplicateFeaturesPerDescJ = duplicateFeaturesPerView[J];
+
+        // per descType
+        for(const auto& matchesIt : matchesPerDesc)
+        {
+            const feature::EImageDescriberType descType = matchesIt.first;
+            const IndMatches& matches = matchesIt.second;
+
+            auto& featuresI = featuresPerDescI.at(descType);
+            auto& featuresJ = featuresPerDescJ.at(descType);
+            auto& duplicateFeaturesI = duplicateFeaturesPerDescI[descType];
+            auto& duplicateFeaturesJ = duplicateFeaturesPerDescJ[descType];
+
+            // per features match
+            for(const IndMatch& m : matches)
+            {
+                {
+                    auto& featureI = featuresI[m._i];
+                    auto& [duplicateFeatureIdsI, duplicateFeatureNodeI] =
+                        duplicateFeaturesI[DuplicateFeatureId(featureI.x(), featureI.y(), featureI.scale())];
+                    IndexedFeaturePair pairI(I, KeypointId(descType, m._i));
+                    auto& nodeI = map_indexToNode.at(pairI);
```

Comment:
In my opinion, the code is too much "implicit" just to avoid some lookup. Moreover, as you choose not to type your variables, that's a little bit cryptic to an external reader.

Reply:
I see. I'll work on improving the code clarity and fix this issue with another commit. Thanks for your feedback.